cocos/2d/CCFontFNT.cpp

```diff
@@ -641,10 +701,22 @@ FontAtlas * FontFNT::createFontAtlas()
         
         tempDefinition.offsetX  = fontDef.xOffset;
         tempDefinition.offsetY  = fontDef.yOffset;
-        
-        tempDefinition.U        = tempRect.origin.x + _imageOffset.x;
-        tempDefinition.V        = tempRect.origin.y + _imageOffset.y;
-        
+
+        const float left = _imageRectInPoints.origin.x;
+        const float right = (_imageRectInPoints.origin.x + rw);
+        const float top = _imageRectInPoints.origin.y;
+
+        if (_imageRotated)
+        {
+            tempDefinition.U = right - tempRect.origin.y - tempRect.size.height;
+            tempDefinition.V = tempRect.origin.x + top;
```

Comment:
Don't understand the codes. Could you please explain it?

Reply:
And i think the variable name is confused.

Reply:
That code is directly out of CCSprite.cpp.  It calculates the boundaries of an image inside the sprite sheet, as you can see in `Sprite::setTextureCoords()`:

```
    float left    = rectInPixels.origin.x / atlasWidth;
    float right   = (rectInPixels.origin.x + rw) / atlasWidth;
    float top     = rectInPixels.origin.y / atlasHeight;
    float bottom  = (rectInPixels.origin.y + rh) / atlasHeight;
```

Now, with the new code in `FontFNT::createFontAtlas()`:

```
        if (_imageRotated)
        {
            tempDefinition.U = right - tempRect.origin.y - tempRect.size.height;
            tempDefinition.V = tempRect.origin.x + top;
        }
        else
        {
            tempDefinition.U = tempRect.origin.x + left;
            tempDefinition.V = tempRect.origin.y + top;
        }
```

If the bitmap font image is rotated 90 degrees to the right in the sprite sheet, then The U and V coordinates need to point to the correct position in the sprite sheet.  

The U and V calculated there result in values that the BM font code expects to extract the BM font image.  Took me a while to figure out what the BM font code required, and that is the correct formula.

Reply:
What i confused is that
- right is calculated with origin.x 
- right is used together with origin.y and size.height

No matter it is rotated or not, it is confused to use origin.x and origin.y together.

Reply:
- float right   = (rectInPixels.origin.x + rw) / atlasWidth;
- tempDefinition.U = right - tempRect.origin.y - tempRect.size.height;

so 
- tempDefinition.U = (rectInPixels.origin.x + rw) / atlasWidth - tempRect.origin.y - tempRect.size.height

The logic is strange.

Reply:
Oh that's because `U` is expected to be the top left position of the letter.  

So, if it's rotated 90° to the right in the sprite, that formula gets the correct coordinate for the top-left of the letter when it's rotated.

Reply:
So what's the coordinate of `_imageRectInPoints ` and `tempRect `:
- does _imageRectInPoints's coordinated based on rotated coordinate or not?
- does tempRect' coordinate based on rotated coordinate or not?

Reply:
`_imageRectInPoints` is the rectangle of the entire bitmap font texture, ignoring rotation.  It's what comes from either a Vec2 imageOffset supplied by the user, or a rect supplied by the user, or the SpriteFrame rect.

`tempRect` is the letter definition as it comes from the FNT file:
```
        BMFontDef& fontDef = e.second;
        ...
        Rect tempRect;
        
        tempRect = fontDef.rect;
        tempRect = CC_RECT_PIXELS_TO_POINTS(tempRect);
```

Example from FNT file:
```
char id=32   x=251   y=31    width=3     height=1     xoffset=-1    yoffset=45    xadvance=11    page=0  chnl=15
```

Reply:
```
float right = (rectInPixels.origin.x + rw) / atlasWidth;
tempDefinition.U = right - tempRect.origin.y - tempRect.size.height
```

That is to calculate the top-left position of the letter image in the bitmap texture.  So, for example:

![image](https://user-images.githubusercontent.com/8603230/68643677-838bc600-0566-11ea-8deb-80fa3fdc5027.png)

```
letter "M" top left  = U = right X position of texture - tempRect.origin.y (green line) - tempRect.size.height (blue line)

and 

V = top + tempRect.origin.x (which is shown by the red line)
```

When it passes this information on to create the Sprite from these coordinates, the sprite can then render the letter out correctly, since it knows it's been rotated.

Reply:
I found rotated information is only used here. How does it affect the frame rotation?

Reply:
Where it's used was the previous PR #20300 

In CCLabel.cpp from that PR, I changed it to use the rotated flag:

Line 793:
`letterSprite->setTextureRect(uvRect, letterDef.rotated, uvRect.size);`

Line 974:
`_reusedLetter->setTextureRect(_reusedRect, letterDef.rotated, _reusedRect.size);`

Line 1802:
`letter = LabelLetter::createWithTexture(_fontAtlas->getTexture(textureID), uvRect, letterDef.rotated);`